### PR TITLE
add custom chunking logic

### PR DESCRIPTION
**Pull request by jotix16**
This commit tries to add custom chunking logic without breaking anything and solve #375. It is related to #376, but doesn't offer a solution for that, no reorganization, it just makes it more complicated.

The `iterate_seqs` gets now an extra input `chunking_logic` that provides with the extra information required. 

#### Inputs 
- sequence(`numpy.array`)
- chunks_step(`int`)
- chunk_size(`int`)
#### Outputs
- chunks_length(`int`): number of chunks that can be generated from given sequence, chunk_size and chunk_step
- custom_chunk_generators(`dict[str, ()->(int, int)]`): dict with generators for every data_key that has a custom chunking logic

The names are not final and can be changed. This commit only presents the logic. Will add Andre's RNNT chunking and some tests afterwards. 

@albert is the idea ok in general?

**Edit**: Is there any script to run the tests locally? Couldn't find any information in the wiki or contribution. 
**Edit**: I am not very acquainted with the tests, but looks like chunking is being tested as part of different tests. Should I create a new file for testing chunking?

### CODE REVIEW

**Comment by albertz** [comment]
This is supposed to be an alternative for #375?

This is still too specific. And also too complicated. And also not documented. Please provide documentation first, before you even implement it. It took a while to understand what you do. Before you even implement sth, please propose the idea first (in pseudo code if you think that is more helpful).

I don't think `chunking_logic` should be a dict. I think it is more generic and simpler to just allow the user to define an own custom `iterate_seqs` function somehow.

**Comment by jotix16** [comment]
>This is supposed to be an alternative for #375? And an implementation for #376?

It is only an alternative for #375. I said it is only related to #376(but doesn't solve anything)


>It took a while to understand what you do. Before you even implement sth, please propose the idea first

The main comment holds the logic.
The chunking logic is an extension to the tuple (chunk_size, chunk_step). Now we have (chunk_size, chunk_step, chunking_logic).

It is a dict because you have to specify the data_key for which the chunking is implemented for.

Why do you think it is too specific?
You can just implement the logic however you wish. I dont think one requires more information than than one sequence and chunk_size, chunk_step to properly specify a chunking logic (or I am not thinking broader than transducer based models)

**Comment by albertz** [comment]
> The main comment holds the logic.

Ah, I did not really read that. You should not explain such things in the comment. This should be documented in the code, docstrings, documentation. No-one will later check the comments in a PR to understand why sth was done this way.

> Why do you think it is too specific?

E.g. your assumptions:

- it is separate per data key (not all approaches might fit into this)
- it gets chunk size / chunk step (that might not really be relevant for more custom stuff)
- it returns chunk_step and chunk_size (not all approaches might fit into this)

This all is very close to the existing approach. What if you (or the user) would want to implement some completely different approach? The your generic function is not generic enough and we again need to have a second more generic function. Which is sth we should avoid.

There should not be any such assumptions. When we already introduce a custom function for the user config, it should be generic. E.g. the user could define a generic `iterate_seqs` function (separate for training and testing).

We might simplify it a bit and still keep it generic. E.g. I think this part of the loop can always be done by us:
```
def seq_idx_generator(self):
    seq_idx = 0
    while self.is_less_than_num_seqs(seq_idx):
       yield seq_idx
       seq_idx += 1
```

And the user defines a function `(Dataset, Iterable[int]) -> Iterable[Tuple[int, NumbersDict, NumbersDict]]` which gets the dataset + `seq_idx_generator()` and returns the same as `iterate_seqs`.

It might also be even more generic if the user directly defines the whole batching logic. E.g. currently this is still not generic enough such that the user could e.g. define/implement the bucketing logic as it is common in other frameworks.

Actually all of that is already possible with our alternative `tf.data` pipeline. So we maybe should also not over-engineer this now, as this is anyway all kind of deprecated.

But all of this discussion is exactly #376.

---

> I am not very acquainted with the tests, but looks like chunking is being tested as part of different tests. Should I create a new file for testing chunking?

There are tests for the dataset, and this is part of the dataset. No new file is needed.

---

Off topic here (use discussion forum for it, or open separate issue for missing documentation or so):

> Is there any script to run the tests locally? Couldn't find any information in the wiki or contribution.

You just run them directly. Or via `nosetests`. (You can also simply look at how GitHub CI runs them.)

**Comment by jotix16** [comment]
``` python
def my_chunking(data, chunk_size, chunk_step):
  blank_indices = np.where(data == blank_index)[0]  # [T]
  blank_length = len(blank_indices)
  len = (blank_length - chunk_size) / chunk_step

  def chunk_generator():
  # ...
    yield chunk_step, chunk_size  

  return len, chunk_generator



chunking = (
   {"data": 12 * 2, "alignment": 22},  # chunk_size
   {"data": 12 * 2 // 2, "alignment": 22 // 2},  # chunk_step
   {"alignment": my_chunking}   # data has the default chunking logic, alignment has a custom one.
  )
```

> This all is very close to the existing approach.

Yes, but it doesn't have the problem that you can't change it anymore. The one you mentioned in #375 : "*But even that is hacky and ugly, and will break in some cases, e.g. when you define custom_iterate_seqs in some epochs, and later not anymore. Then it would not correctly reset this.*"

> We might simplify it a bit and still keep it generic. E.g. I think this part of the loop can always be done by us:

You are right, this looks much more flexible. But I don't really like the idea that it should be dataset dependent. Chunking is a simple thing, it shouldn't require the whole dataset instance to be done. It gets more difficult for the user to implement a custom_iterator, i.e. is not straightforward without a previous example.

Could you mention one such scenario when such a complicated chunking is required? Maybe then I can better understand what I am missing.

**Comment by albertz** [comment]
> This is a dummy usage in case I didn't make it clear: ...

Note: Before you implement sth, this should exactly be the proposal, and then to be discussed.

When you implement sth and make a PR, this should not be a comment, but part of the code / docstring / documentation.

> > This all is very close to the existing approach.
> 
> Yes, but it doesn't have the problem that you can't change it anymore. The one you mentioned in #375 : "_But even that is hacky and ugly, and will break in some cases, e.g. when you define custom_iterate_seqs in some epochs, and later not anymore. Then it would not correctly reset this._"

This comment was purely about the specific implementation in the #375 PR, not about the generic approach to provide a custom `iterate_seqs` function.

> > We might simplify it a bit and still keep it generic. E.g. I think this part of the loop can always be done by us:
> 
> You are right, this looks much more flexible. But I don't really like the idea that it should be dataset dependent. Chunking is a simple thing, it shouldn't require the whole dataset instance to be done. It gets more difficult for the user to implement a custom_iterator,

I don't exactly understand. This is not dataset dependent?

> i.e. is not straightforward without a previous example.

I see this argument. But this is a problem with almost all solutions. Also for your case, a user would not really be able to do this without having seen such an example before.

So we just give this example. The example for such a custom iterator is just as simple/short, or maybe even shorter as your example.

> Could you mention one such scenario when such a complicated chunking is required? Maybe then I can better understand what I am missing.

E.g. to implement the standard batching behavior as it present in most other frameworks (bucketing).

Or to have some logic where you select from `classes` to contain some specific class indices, or be more uniform, and then the `data` accordingly to match the `classes` selection.

Or whatever. Maybe even you could do sort of data augmentation / shuffling or so at this point. Or image translation, where you select sub parts of source and target sentence according to some logic.

Maybe returning `Tuple[int, NumbersDict, NumbersDict]` is also already too specific, and you would just return directly the data (dict data key -> numpy array).

**Comment by jotix16** [comment]
> Note: Before you implement sth, this should exactly be the proposal, and then to be discussed.

Now I know :)

> It might also be even more generic if the user directly defines the whole batching logic. E.g. currently this is still not generic enough such that the user could e.g. define/implement the bucketing logic as it is common in other frameworks.

> E.g. to implement the standard batching behavior as it present in most other frameworks (bucketing).

I am a little confused now. Isn't batching and chunking independent? Batching comes after chunking in the pipeline.
This flexibility you are saying involves only the batching logic in '_generate_batch`. 

**batching** is in sequence level
**chunking** is in frame level

```
      seqs                   seqs'                     batches
     ----->     chunking     ----->     batching      --------->
```  

is this right?

**Comment by albertz** [comment]
> I am a little confused now. Isn't batching and chunking independent? Batching comes after chunking in the pipeline.
> This flexibility you are saying involves only the batching logic in '_generate_batch`.

They are in our pipeline. But chunking is just one custom post processing thing you can do (actually not so common, I'm not sure you find any other framework having this). There are many other potential post processing steps you could do. The general interface to allow any kind of post processing would be some function seqs -> seqs (output: yielding dict with data key -> numpy array).

Sure you can always keep batching as a separate step. Sorry, my suggestion was probably confusing. I thought you could also generalize this part just in the same way. And there could be some logic where you want to combine the pipeline. E.g. prepare chunks exactly such that it is optimal for batching or so.

> **batching** is in sequence level
> **chunking** is in frame level
> 
> ```
>       seqs                   seqs'                     batches
>      ----->     chunking     ----->     batching      --------->
> ```
> 
> is this right?

Yes, this is our custom pipeline (when we do not use `tf.data`).

Note that `tf.data` and our support in RETURNN is very much this generic pipeline already. The user just provides the final batched `tf.data.Dataset`.

I think it is also similar in PyTorch.

**Comment by albertz** [comment]
Obsolete via #1075.
